### PR TITLE
Add particle filter in ParticleReductionFunctor

### PR DESCRIPTION
**Pull request by hklion**
This adds an optional filter function to the particle averaging diagnostic (#2892). Particles are included or not based on a filter function that takes `(x, y, z, ux, uy, uz)`. The `particle_fields_diags` tests also test the filter for both plotfiles and openPMD.

### CODE REVIEW

File: Source/Diagnostics/ComputeDiagFunctors/ParticleReductionFunctor.H [hunk 1]
```diff
@@ -41,10 +45,15 @@ public:
 private:
     int const m_lev; /**< level on which mf_src is defined */
     int const m_ispec; /**< index of species to average over */
+    bool const m_do_filter; /**< whether to apply #m_filter_fn */
```
Comment (ax3l) [inline Source/Diagnostics/ComputeDiagFunctors/ParticleReductionFunctor.H:48]: Is the `#` intentional?
```suggestion
    bool const m_do_filter; /**< whether to apply m_filter_fn */
```

File: Source/Diagnostics/Diagnostics.H [hunk 2]
```diff
@@ -155,6 +155,10 @@ protected:
     std::vector< int > m_pfield_species_index;
     /** Map connecting particle field variable names to the corresponding parser strings */
     std::map< std::string, std::string > m_pfield_strings = {};
+    /** Whether to use a filter function on particles before calculating particle averages */
+    std::map< std::string, bool> m_pfield_dofilter = {};
```
Comment (ax3l) [inline Source/Diagnostics/Diagnostics.H:160]: Just realized that we should include `<map>` in this file. Can you add this pls?

**Review by ax3l — COMMENTED** [review]
This is all looking excellent, two minor comments added.

File: Source/Diagnostics/ComputeDiagFunctors/ParticleReductionFunctor.H (hunk 1 above)
Comment (hklion) [reply Source/Diagnostics/ComputeDiagFunctors/ParticleReductionFunctor.H:48]: The `#` in the doxygen comment links to the member variable, e.g. 
https://warpx.readthedocs.io/en/latest/_static/doxyhtml/class_particle_reduction_functor.html#af0e36574686a5b6163543f31a2ff43b5

It's definitely not necessary, though.

File: Source/Diagnostics/Diagnostics.H (hunk 2 above)
Comment (hklion) [reply Source/Diagnostics/Diagnostics.H:160]: Good catch. Done.

File: Source/Diagnostics/ComputeDiagFunctors/ParticleReductionFunctor.cpp [hunk 3]
```diff
@@ -81,7 +90,9 @@ ParticleReductionFunctor::operator() (amrex::MultiFab& mf_dst, const int dcomp,
                 amrex::ParticleReal ux = p.rdata(PIdx::ux) / PhysConst::c;
                 amrex::ParticleReal uy = p.rdata(PIdx::uy) / PhysConst::c;
                 amrex::ParticleReal uz = p.rdata(PIdx::uz) / PhysConst::c;
-                amrex::Real value = map_fn(xw, yw, zw, ux, uy, uz);
+                amrex::Real value;
+                if ((do_filter) && (filter_fn(xw, yw, zw, ux, uy, uz) <= 0.5)) value = 0._rt;
```
Comment (hklion) [inline Source/Diagnostics/ComputeDiagFunctors/ParticleReductionFunctor.cpp]: If #3007 is merged, this condition will need to be changed for consistency, as will the one on line 133.

File: Source/Diagnostics/ComputeDiagFunctors/ParticleReductionFunctor.H (hunk 1 above)
Comment (ax3l) [reply Source/Diagnostics/ComputeDiagFunctors/ParticleReductionFunctor.H:48]: Oh cool, I thought it magically maps them by name. TIL! :)

File: Source/Diagnostics/ComputeDiagFunctors/ParticleReductionFunctor.cpp (hunk 3 above)
Comment (ax3l) [reply Source/Diagnostics/ComputeDiagFunctors/ParticleReductionFunctor.cpp]: Ok, merged :) Ready for update :)

File: Source/Diagnostics/ComputeDiagFunctors/ParticleReductionFunctor.cpp [hunk 4]
```diff
@@ -45,8 +52,10 @@ ParticleReductionFunctor::operator() (amrex::MultiFab& mf_dst, const int dcomp,
     red_mf.setVal(0._rt);
     ppc_mf.setVal(0._rt);
     auto& pc = warpx.GetPartContainer().GetParticleContainer(m_ispec);
-    // Copy over compiled parser function so that it can be captured by value in the lambda
+    // Copy over member variables so they can be captured in the lambda
     auto map_fn = m_map_fn;
+    auto filter_fn = m_filter_fn;
+    bool do_filter = m_do_filter;
```
Comment (EZoni) [inline Source/Diagnostics/ComputeDiagFunctors/ParticleReductionFunctor.cpp]:
```suggestion
    const bool do_filter = m_do_filter;
```

**Review by EZoni — APPROVED** [review]
Thank you for this PR, Hannah! Looks good to me. I just left one suggestion for a boolean temporary variable that can probably be declared as `const`, if I didn't overlook the code.

**Review by ax3l — APPROVED** [review]
Thank you! :tada: 

I will merge, NVHPC has some temporary network problem.

**Pull request merged**